### PR TITLE
Add 'never' type

### LINKED ISSUE #3076: Special `bottom` type
I think that it should be special **bottom** type with the following properties:

* ( T | **bottom** ) = ( **bottom** | T ) = T   — **bottom** is unit for type unions;
* expression of type **bottom** is assignable to any other type;
* **bottom** type could be used only as return type of the function (out-only);
* if function has return type **bottom** then  all branches of the control flow either throw exceptions or return expression of type **bottom**;
*  if all branches of the control flow either throw exceptions or return expression of type **bottom** and function has no explicit return type then **bottom** type inferred as the result type;
* ~~two functional types are not assignable to each other if exactly one of them has return type **bottom**~~ (?);
* **bottom** type can't be used as the actual type parameter for the generic type.

### rationale

Let's consider the function that always throws:

```TypeScript
function alwaysThrows() {
    throw new Error("!");
}
```

and other function that uses it

```TypeScript
function test(a: boolean) {
	return a ? 123 : alwaysThrow();
}
```

We expect that `test` should have type `number` instead of `( number | void )`

In more sophisticated example that involves Promise type declarations:

```TypeScript
declare class Promise<T> {
	catch<U>(
		handler: (reason: any) => U | Promise<U>
	): Promise<T | U>;  // Note the union here (!)
	then<U>(
		fulfilled?: (result: T) => U | Promise<U>,
		rejected?: (reason: any) => U | Promise<U>
	): Promise<U>
}
```

We expect that

```JavaScript
promiseOfString.catch(function(e) {
   throw new BusinessError(e);
});
```

should have the type of `Promise<string>`  instead of  `Promise<string | void>`.

See https://github.com/petkaantonov/bluebird/issues/589#issuecomment-99979305

Possible explicit annotation that function always throws is using **undefined** as the type

```TypeScript
function alwaysThrows(): undefined {
    throw new Error("!");
}
```

### LINKED ISSUE #8602: Consider loosening strict null checks & implicit return checking in the presence of the 'nothing' type
From a discussion with @wycats.

Currently, even if a type is narrowed down to `nothing`, we still consider control flow errors in those branches even though you might just be handling error cases:

```ts
// Error: Not all code paths return a value.
function foo(x: string | number) {
    if (typeof x === "string") {
        return true;
    }
    else if (typeof x === "number") {
        return false;
    }
    fail("Unexhaustive!");
}

function fail(message: string) {
    throw message;
}
```

Similarly, if you add a type annotation, you'll get an error.

```ts
// Error: Function lacks ending return statement but return type lacks 'undefined'
function foo(x: string | number): boolean {
    if (typeof x === "string") {
        return true;
    }
    else if (typeof x === "number") {
        return false;
    }
    fail("Unexhaustive!");
}

function fail(message: string) {
    throw message;
}
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR introduces a `never` type that represents the type of values that never occur. Specifically, `never` is the return type for functions that never return and `never` is the type of variables under type guards that are never true. The `never` type has the following characteristics:

* `never` is a subtype of and assignable to every type.
* No type is a subtype of or assignable to `never` (except `never` itself).
* In a function with no return type annotation, if the function has no `return` statements, or only `return` statements with expressions of type `never`, and if the end point of the function is not reachable (as determined by control flow analysis), the inferred return type for the function is `never`.
* In a function with an explicit `never` return type annotation, all `return` statements (if any) must have expressions of type `never` and the end point of the function must not be reachable.

Because `never` is a subtype of every type, it is always omitted from union types and it is ignored in function return type inference as long as there are other types being returned.

The `never` type replaces the `nothing` type introduced in #8340 and it is effectively the same as the `bottom` type proposed in #3076.

Some examples of functions returning `never`:

```typescript
// Inferred return type is never
function error(message: string) {
    throw new Error(message);
}

// Inferred return type is never
function fail() {
    return error("Something failed");
}

// Function returning never must have unreachable end point
function infiniteLoop(): never {
    while (true) {
    }
}
```

Some examples of use of functions returning `never`:

```typescript
// Inferred return type is number
function move1(direction: "up" | "down") {
    switch (direction) {
        case "up":
            return 1;
        case "down":
            return -1; 
    }
    return error("Should never get here");
}

// Inferred return type is number
function move2(direction: "up" | "down") {
    return direction === "up" ? 1 :
        direction === "down" ? -1 :
        error("Should never get here");
}

// Inferred return type is T
function check<T>(x: T | undefined) {
    return x || error("Undefined value");
}
```

Because `never` is assignable to every type, a function returning `never` can be used when a callback returning a more specific type is required:

```typescript
function test(cb: () => string) {
    let s = cb();
    return s;
}

test(() => "hello");
test(() => fail());
test(() => { throw new Error(); })
```

Fixes #3076.
Fixes #8602.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
I really don't think this name is good for the general user experience. My feeling is that it seems to model several different things which the name doesn't reflect very well on. At least `nothing` reflected the domain of values a little bit better.

**Comment by ahejlsberg** [comment]
@DanielRosenwasser The primary use for `never` is as the return type of functions that never return. I'm hard pressed to come up with a more descriptive name. Certainly `nothing` isn't in my mind.

**Comment by RyanCavanaugh** [comment]
The problem with `nothing` is that it's very confusing to then say "`void` is the return type of functions that return nothing". `never` also parses well with quick info in typeguarded code:

```ts
function f(x: A | B) {
  if(isA(x)) {
  } else if(isB(x)) {
  } else {
    x; // appears as type 'never'
  }
}
```

**Comment by sandersn** [comment]
My objection to `never` is that it reads strangely: "`f` returns `never`" wants to be "`f` never returns". But "`f` returns `nothing`" reads fine.
It would help to show some examples of `never` actually used in some of the tests.

Other than that, and bike shedding on the name, the code looks good.

**Comment by DanielRosenwasser** [comment]
Agreed with Nathan - the code looks fine, but I have those gripes with the naming.

**Comment by ahejlsberg** [comment]
Yes, but "`f` returns `nothing`" also strongly implies that `f` actually returns. Which is precisely the intuition we want to avoid!

**Comment by sandersn** [comment]
Interesting. That makes it sounds intended that none of the tests explicitly mention their return type is `never`. Could we get away with not making `never` a keyword and having it always be inferred?

**Comment by ahejlsberg** [comment]
No, we can't get away with `never` always being inferred. We need it in declaration files, for example. Several of the tests in the PR already explicitly specify `never` as their return type.

**Comment by sandersn** [comment]
OK, that makes sense. The d.ts is a good enough reason to have the keyword, although I noticed the rest -- outside a d.ts -- are all error cases where the annotation specifies `: never` and then fails to get it.

**Comment by mhegazy** [comment]
This will be a breaking change for back compat for .d.ts emit. in the past we inferred the type of these methods/functions as `void`. some one using nightly builds will generate .d.ts files that have the new type `never` that is not consumable by TS 1.8 users.

**Comment by ahejlsberg** [comment]
@mhegazy Yes, but an easy fix is to add a `void` type annotation.

**Comment by mhegazy** [comment]
we said we will be more careful in the future with such changes.

**Comment by ahejlsberg** [comment]
I don't see a big issue here. We're being more precise in our type analysis and that may change the outcome of inference. There are other such issues already resulting from the control flow based type analysis (e.g. narrowing occurs in more places which might affect inferred return types). Effectively we can only replicate the old results by keeping the old code and not changing anything. I think it is reasonable to recommend explicit type annotations if you want to freeze the shape of an API.

**Comment by mhegazy** [comment]
Being more precise is one issue. Generating a .d.ts that is not consumable is another.

**Comment by ahejlsberg** [comment]
I'm not sure where you're going with this. What are you proposing?

**Comment by yortus** [comment]
@ahejlsberg I've just submitted an issue (#8655) about better type analysis around `assert(...)` calls. Could this `never` type be a step toward that kind of analysis becoming possible in future? `assert` functions might require a return type annotation like `void | never` or `x is T | never` to tell consumers they either prove a type assertion or don't return at all.

**Comment by aleksey-bykov** [comment]
Sorry for not minding my own business. The official 1.8. has its *.d.ts fixed and shipped, whoever needs it is welcome to get it anytime. Technically speaking nightly build is 1.9.+ with a bump in the major version number which means breaking changes are likely. Can't see any crime here. People who are scared to break their code should upgrade are warned to proceed with caution or stay where they are. Am I missing the point?

**Comment by kitsonk** [comment]
> The official 1.8. has its *.d.ts fixed and shipped, whoever needs it is welcome to get it anytime. Technically speaking nightly build is 1.9.+ with a bump in the major version number which means breaking changes are likely. Can't see any crime here. People who are scared to break their code are warned to proceed with caution or stay where they are. Am I missing the point?

It isn't only the TypeScript team that generates `.d.ts` files. 😄 Meaning that there is a fracture in the entire ecosystem.  For example if Angular 2, Dojo 2, RxJS, etc. migrate then it causes them to generate definition files that are not consumable by those that haven't upgraded, continuing to fracture and annoy the wider TypeScript community.  Many large entities don't have the time or resources to keep upgrading every 6 weeks.

Both @ahejlsberg and @mhegazy have points.  My opinion is if this lands for 2.0, there is the `readonly` breaking change.  Arguably that is opt-in, while in theory, this would likely end up not being opt-in, but invariably people are going to start producing `.d.ts` files with `readonly` anyways and stranding < 2.0 users.  So I would say 2.0 would be the point to break things from an interface perspective.

**Comment by RyanCavanaugh** [comment]
We've actually broken downlevel .d.ts consumption with almost every release. Union types were inferable, as were intersection types once upstream libraries started using them in their definitions.

It's valid to add a `: void` type annotation to any function which has an inferred type of `never`. People who want to generate downlevel-compatible .d.ts will always need to take precautions regardless of what we do; checking for `never` annotations and resurfacing them as `void` is just another item on that checklist.

Generating `: void` in lieu of `: never` in a .d.ts until 2.0 seems like a reasonable compromise, though.

**Comment by Artazor** [comment]
Cool! Now #3076 can be closed!

**Comment by aleksey-bykov** [comment]
> For example if Angular 2, Dojo 2, RxJS, etc. 

Not quite seeing the issue in the example. Teams like Dojo 2, etc. can either:

- follow the release cycle of TS thus staying always in sync
- commit to support both the previous and the current version of TS if their release cycle is misaligned
- care less by disclaiming on what TS version is supported
- do whatever their engineer's intuition tells them for this is what engineering is about - solving the technical problems, especially ones well-studied like this: maintaining compatibility between the components

Bottom line is that if a new feature is bound to break the existing code it will break it sooner or later. The only part that is worth discussion when and how it will happen.

**Comment by kitsonk** [comment]
> Not quite seeing the issue in the example.

These people (for disclosure, I am the lead for Dojo 2) issue `.d.ts` which are then consumed by other people to build applications.  Your original statement was:

> The official 1.8. has its *.d.ts fixed and shipped, whoever needs it is welcome to get it anytime.

But instead of acknowledging that there is potentially more breakage than just what version of the `lib.d.ts` someone is using, you suggest that project like Dojo 2 (of which I am the lead) can start issuing versions of their `.d.ts` files without regard for the fact that these projects have a downstream user base.  There is a risk with a non-optional breakage like this that we could have serious impacts downstream.

I clearly indicated that both Anders and Mohammad had valid points, of which you seemed to suggest in your original statement confusion about why Mohammad was concerned.  I am glad he is concerned.  I know both Anders and Mohammad consider downstream projects like Dojo 2 and others in their decisions.  Ryan further added though that valid point that this isn't the first breakage of definition files that force the larger community to upgrade.  I would argue though that intersection and union types were opt-in, so in theory downstream projects could have chosen not to use them, but they were actually key enablers and so I know we chose to opt-in to them.

Clearly the compromise of emitting a `void` in declaration files would be perfect and cause a whole lot less downstream carnage.  I am really glad to know that the team thinks about these things instead of going "oh, whats the big deal, people can just upgrade"...

**Comment by aleksey-bykov** [comment]
>you suggest that project like Dojo 2 (of which I am the lead) can start issuing versions of their .d.ts files

Where did I say that?

What I said is: it's a purely engineering problem, and this is what we, engineers, do for living: solve them, and not to seem too arrogant I even outlined how it can be done in typical cases.

Now if the budget of TS is big enough to take on supporting backward compatibility for every team that might benefit of it, I am all for it. Doubtfully so.

More realistically the budget of TS can be better spent on developing new features merits of which far outweigh the troubles of adopting them. And this is what I am rooting for.

Lastly please give an example how `v.2.0/lib.d.ts` has anything to do with `v.1.8/dojo.d.ts`?

**Comment by ahejlsberg** [comment]
I'm not sure generating `: void` instead of `: never` in .d.ts files until we ship 2.0 will materially change anything. Aside from appearing to be a bug, it just delays the realization that something has changed. In fact, if we're going to break something then we should break it early in a nightly build to give folks time to adapt. Introducing the actual break right as we ship seems really bad.

We already have a breaking change in 2.0 with `readonly` because we sometimes infer it (e.g. when you have a get accessor without a set accessor). I think this one is very similar. I also think it is quite reasonable to ship these changes as long as we document them and the appropriate workarounds.

Alternative solutions are:

* Never infer `never` as a return type and instead require an explicit type annotation.
* Introduce (yet another) switch to enable or disable inference of `never` return types.

I think both of those are worse than just introducing the change. There is real value in having inference produce `never` as it calls out an important behavioral distinction. We lose that with an opt-in model. And we really don't want more switches.

**Comment by mhegazy** [comment]
In https://github.com/Microsoft/TypeScript/issues/8252 we said we will add this to our "feature checklist". i am just bringing this up as a violation of this. 

Given that we rejected the sourceVersion suggestion earlier this weeks, I believe there are two options here, the first is the one you mentioned (that also should apply to `undefined` and `null` now that i think about it), and the second is to say this is too much of a restriction, and it is not supported.
My vote would be for the second.

**Comment by ahejlsberg** [comment]
Discussed with @mhegazy. We're good with documenting this as a breaking change for .d.ts emit and will also document the workaround (add `void` type annotations).

**Comment by JsonFreeman** [comment]
What a nice feature! I also quite like the name.

**Comment by mhegazy** [comment]
Note for breaking changes docs:
A common break would be a base class with an implementation to always throws. previously the return was inferred as `void`, now it is `never`. 

Example:

```ts
class Base {
    method() {
        throw new Error("Not Implemented!");
    }
}

class Derived extends Base {
    method() {
        return 0; 
    }
}

// error TS2415: Class 'Derived' incorrectly extends base class 'Base'.
//   Types of property 'method' are incompatible.
//     Type '() => number' is not assignable to type '() => never'.
//       Type 'number' is not assignable to type 'never'.
```

Recommendation:

Give the base class method a type annotation of the expected return type, be it `void` or any other concrete type.

**Comment by ahejlsberg** [comment]
With #8767 we have modified the inference rules to fix the breaking change above.

**Comment by usta** [comment]
how about naming it as "impossible" instead of "never"

**Comment by kostat** [comment]
I understand that the primary reason to call the type "never" derives from the desire to facilitate the reading the never returning function declaration. For me, it does not reflect the essence of the type. Look:

`var x:never = IAlwaysThrow();`

On the other hand the common definition of such code is unreachable code block. Wouldn't it be better to name the type "unreachable"?

```
function IAlwaysThrow() : unreachable {
    throw Error();
}

var x:unreachable = IAlwaysThrow();
```

**Comment by kitsonk** [comment]
Considering this is already merged and documented, all we are doing is 🚲 🏠 and navel gazing something is a semantic opinion.  The naming was already debated and settled.

**Comment by fightingcat** [comment]
What's the point?

**Comment by kitsonk** [comment]
@fightingcat the point is that TypeScript lacked a [bottom type](https://en.wikipedia.org/wiki/Bottom_type) and that made it difficult to properly enforce parts of a type system.

**Comment by remojansen** [comment]
Can someone please explain me why given the following code:

```ts
let f = () => {
    throw new Error("Should never get here");
}

let g = function() { 
    throw new Error("Should never get here");
}

function h() { 
    throw new Error("Should never get here");
}
```

The following types are inferred:

- `f` is `() => never`
- `g` is `() => never`
- `h` is `() => void`

Thanks!

**Comment by RyanCavanaugh** [comment]
@remojansen please use Stack Overflow for questions -- thanks!

**Comment by remojansen** [comment]
@RyanCavanaugh sorry my mistake will move it to SO.
